File: 377.组合总和IV.cpp
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        if(nums.size() == 0)
            return 0;
        if(target == 0)
            return 1;
        sort(nums.begin(), nums.end());
        unsigned long long* dp = new unsigned long long[target + 1]();
        dp[0] = 1;
        for(int i = 1; i <= target; ++i){
            for(int j = 0; j < nums.size() && nums[j] <= i; ++j)
                dp[i] += dp[i - nums[j]];
        }
        return dp[target];
    }
};

File: 动态规划dp/377.组合总和IV.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var combinationSum4 = function (nums, target) {
    let dp = new Array(target + 1).fill(0);
    nums = nums.sort((a, b) => {
        return a - b;
    });
    dp[0] = 1;
    for (let i = 1; i <= target; ++i) {
        for (let j = 0; j < nums.length && nums[j] <= i; ++j) {
            dp[i] += dp[i - nums[j]]
        }
    }
    return dp[target];
};